### PR TITLE
Ampers: Kirsten and Kaitlin

### PR DESCRIPTION
**Pull request by kschumy**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? |  It allows you to share and edit each other's work from separate computers. 
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | So, instead of creating new instance of score for each word,  you can use the Class method directly on the word and store it value as local variable. 
Describe an example of a test you wrote for a _nominal case_. |  We can instantiate a TileBag class
Describe an example of a test you wrote for an _edge case_. | Testing if the tilebag decreased with letters drawn but not beyond the bag's inventory
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used group_by to get the highest score and its word. It reduced the code to do the iteration
What was one method you and your pair used to debug code? | We used pry in tile bag to see what was being removed because the count wasn't changing correctly.
Is there a specific piece of code you'd like feedback on? | The tests and what are necessary tests
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | Feedback on the code itself that allowed us to update and edit it without getting overprotective of it

### CODE REVIEW

File: lib/dictionary.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    def self.is_word?(word)
+      raise ArgumentError.new("Invalid input: #{word}") if word.class != String
+      build_all_words if @@all_words.empty?
+      return @@all_words.include?(word)
+    end
+
+    private
+
+    def self.build_all_words
+      words = CSV.read('support/all_words.csv')
```
Comment (CheezItMan) [inline lib/dictionary.rb]: Your support folder isn't checked into github

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,81 @@
...
+      end
+    end
+
+    # Throws ArgumentError if provided word is not an String, is not between 1
+    # and MAX_LETTERS (inclusive) in size, or contains characters other than
+    # a-z/A-Z.
+    def self.has_valid_initial_word?(word)
+      if word.class != String || !word.size.between?(1, MAX_LETTERS) ||
+        !word.upcase.match?(/^[A-Z]+$/)
+        raise ArgumentError.new("#{word} is not a word!")
```
Comment (CheezItMan) [inline lib/scoring.rb:86]: Not in the requirements, but NICE

File: specs/scoring_spec.rb [hunk 3]
```diff
@@ -25,38 +28,60 @@
...
+    it 'throw ArgumentError if element in provided array is not a String' do
+      proc { Scrabble::Scoring.highest_score_from(["foo", 4, "bar"]) }.must_raise ArgumentError
     end
 
     it 'returns the only word in a length-1 array' do
+      Scrabble::Scoring.highest_score_from(["unicorn"]).must_equal "unicorn"
     end
 
     it 'returns the highest word if there are two words' do
+      Scrabble::Scoring.highest_score_from(["unicorn", "boo"]).must_equal "unicorn"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:70]: You should also try varying the order in these tests, just to ensure it's not the order determining the winner.

File: specs/scoring_spec.rb [hunk 4]
```diff
@@ -25,38 +28,60 @@
...
     end
 
     it 'if tied, prefer a word with 7 letters' do
+      Scrabble::Scoring.highest_score_from(["QQQQQQ", "BEAIOUD"]).must_equal "BEAIOUD"
     end
 
     it 'if tied and no word has 7 letters, prefers the word with fewer letters' do
+        Scrabble::Scoring.highest_score_from(["QQ", "QKK"]).must_equal "QQ"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb]: Good!

File: lib/tilebag.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        tiles << @tiles_in_bag.slice!(rand(0...@tiles_in_bag.size))
+      end
+      @tiles_remaining -= num
+      return tiles
+    end
+
+    private
+
+    # Creates bag of tiles and sets tiles_remaining value.
+    def build_tiles_in_bag
```
Comment (CheezItMan) [inline lib/tilebag.rb:34]: This is great!

File: specs/player_spec.rb [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    it 'stores words' do
+      player = Scrabble::Player.new('BamBam')
+      player.plays.must_equal []
+      player.play("PUDDLE")
+      player.plays.must_equal ["PUDDLE"]
+    end
+
+    it 'stores more than one of the same word' do
+      player = Scrabble::Player.new('BamBam')
+      2.times { player.play("PUDDLE") }
```
Comment (CheezItMan) [inline specs/player_spec.rb:72]: You might try more than the same word, like:
```ruby
["PUDDLE", "DOG"].each do |word|
  player.play(word)
end

["PUDDLE", "DOG"].each do |word|
  player.plays.must_include(word)
end
```

File: specs/player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+    it 'does not add invalid input for word' do
+      player = Scrabble::Player.new('BamBam')
+      proc { player.play(nil) }.must_raise ArgumentError
+      player.plays.include?(nil).must_equal false
+    end
+  end
+
+  describe 'total_score' do
+    it 'returns sum of all players word scores' do
```
Comment (CheezItMan) [inline specs/player_spec.rb:109]: You should also test to ensure that `total_score` doesn't increase **after** the player has won.

File: specs/tilebag_spec.rb [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      tile_bag = Scrabble::TileBag.new
+      tiles_from_bag = tile_bag.tiles_in_bag
+      tiles_from_bag.all?(String).must_equal true
+      tiles_from_bag.all? { |letter| letter.size == 1 }.must_equal true
+      tiles_from_bag.all? { |letter| letter.match?(/[A-Z]/) }.must_equal true
+    end
+
+    it 'set up tiles_remaining at initialize' do
+      tile_bag = Scrabble::TileBag.new
+      tile_bag.tiles_remaining.must_equal 98
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb:34]: This seems to repeat code above.

File: specs/tilebag_spec.rb [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    end
+  end
+
+  describe 'draw_tiles(num)' do
+    it 'returns a collection of random tile' do
+      tile_bag = Scrabble::TileBag.new
+      draw = tile_bag.draw_tiles(5)
+      draw.length.must_equal 5
+      draw.must_be_kind_of Array
+      assert draw.all? { |letter| letter.class == String }
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb:44]: Better to use spec-style testing with `.class.must_equal` rather than `assert`

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Check, testing an attempt to draw **more** tiles than are in the bag would be an _edge-case_
Both Teammates contributed to the codebase |  It appears so, it looks like you just paired on Kirsten's computer so only she appears on the commits.  
Regular Commits with meaningful commit messages |  Good commit messages and **lots** of commits.  
**Scoring class** |
`score` method | Check, well done
Uses appropriate data structure to store the letter score | Nicely done
Appropriately handles edge cases |  Check, see a small note in your code
`highest_score_from` method |  Check
Appropriately handles edge cases |  Check
Test implementations for `highest_score_from`|  Check
**Player class** |
Uses an array to keep track of words played |  Check
Uses existing code to create `highest_scoring_word` and `highest_word_score` | Check
Returns `true` or `false` from `won?` method |  Check
Tests edge cases on the `play(word)` | Check
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag |  Check, nicely done!
`draw_tiles` method uses the parameter to draw a set of random tiles |  Check
`tiles_remaining` method returns a count |  Check
Tests `draw_tiles` edge cases |  Well done
Summary | Very well done, you hit all the learning goals.  I'm just sorry you didn't check the dictionary csv file into github.  Outstanding.

**Comment by kschumy** [comment]
Chris,

Thanks for the feedback! 

Concerning the "both teammates contributed to the codebase" component on the feedback sheet, we passed my computer back and forth for all the non-optional requirements, so that's why it the vest majority of commits are from my account. Doing this helped us adhere to the sharing aspect of pair programming. For the optional requirements, we worked on some things separately and some things together, which is why again mostly are from my account.

Just wanted to make sure Kaitlin get her due credit!

Kirsten
